contracts/sophia_explained.md

```diff
@@ -0,0 +1,111 @@
+But instead of defining multiple `event` structs in Sophia *one* instance of
+the `event` data type is defined - with one type constructor per event we want
+to emit. For example the following definition makes it possible to emit two
+different events `TheFirstEvent` and `AnotherEvent`.
+
+```
+  datatype event =
+      TheFirstEvent(indexed int)
+    | AnotherEvent(indexed address, string)
+```
+
+An event may have 0-3 `indexed` fields, these fields should have a Sophia type
+that is represented as one 32-byte word at the VM level (for example int, bool
+and address). An event may also have an additional non-indexed field, this
+field should be of type `string`. (Note: the `indexed` keyword is currently
+mandatory, there are no non-indexed word arguments, this keyword might be
+removed in later versions of Sophia.) Events are emitted by using the
+`Chain.event` function. The following function will emit one Event of each kind
+in the example.
+
+```
+  public function emit_events() : () =
+    Chain.event(TheFirstEvent(42))
+    Chain.event(AnotherEvent(Contract.address, "This is not indexed"))
+```
+
+### Accessing events
+
+The events are stored in the [contract call
+object](../serializations.md#contract-call) and can be accessed by looking up
+the transaction in which they were emitted. If we assume that `emit_events`,
+from our example, was called in a transaction with transaction hash
+`th_a5239bd8e3...` we can access it from the HTTP API
+`/v2/transactions/th_a5239bd8e3.../info`:
+
+```
+#{<<"caller_id">> => <<"ak_XSnUeXPB8UncW8d5R9xoov8x5iTnA8NrURdqLGn9PnyAXYY78">>,
+  <<"caller_nonce">> => 4,
+  <<"contract_id">> => <<"ct_2pvLLjPfjqRMSX91BgDRK8gLzeJD4qaQtNF5F6GdWKamCCnZc">>,
+  <<"gas_price">> => 1000000000,
+  <<"gas_used">> => 3274,
+  <<"height">> => 7,
+  <<"log">> =>
+      [#{<<"address">> => <<"ct_2pvLLjPfjqRMSX91BgDRK8gLzeJD4qaQtNF5F6GdWKamCCnZc">>,
+         <<"data">> => <<"cb_Xfbg4g==">>,
+         <<"topics">> => [100682519114738024061100211920083630085287628982043971034774722609724357358252,
+                          42]},
+       #{<<"address">> => <<"ct_2pvLLjPfjqRMSX91BgDRK8gLzeJD4qaQtNF5F6GdWKamCCnZc">>,
+         <<"data">> => <<"cb_VGhpcyBpcyBub3QgaW5kZXhlZK+w140=">>,
+         <<"topics">> => [84776946516659401778974935644603388793657850787716942371444147005700991736047,
+                          1875564187002476023854543820981509249331367502514562901623081521315128929137]}],
+  <<"return_type">> => <<"ok">>,
+  <<"return_value">> => <<"cb_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAArMtts">>}
+```
+
+First look at the `data` field. For the first event it is `<<"cb_Xfbg4g==">>`
+which decodes to `<<>>`, i.e. no data (and that is expected since the event did
+not have any extra data.). For the second event we see
+`<<"cb_VGhpcyBpcyBub3QgaW5kZXhlZK+w140=">>` which decodes to `<<"This is not
+indexed">>`. For the `topics` field we see that there is actually two values
```

Comment:
I think that the first value of the `topics` field can be explained a bit more. For the first example event the `indexed int` value is presented as a plain integer (42), but for the second event the `indexed address` value is, I assume, hash of the address. The logic behind this is not explained, what would be the value if the parameter is a `boolean`?